src/core/xconnection.rs

```diff
@@ -53,62 +53,85 @@ const EVENT_MASK: &[(u32, u32)] = &[(
         | xcb::EVENT_MASK_SUBSTRUCTURE_NOTIFY,
 )];
 
-// TODO: this list has been copied from atoms used in other WMs, not using everything
-//       yet so work out which ones we need to keep and which we can drop.
-const ATOMS: &[&'static str] = &[
-    "MANAGER",
-    "UTF8_STRING",
-    "WM_CLASS",
-    "WM_DELETE_WINDOW",
-    "WM_PROTOCOLS",
-    "WM_STATE",
-    "WM_NAME",
-    "WM_TAKE_FOCUS",
-    "_NET_ACTIVE_WINDOW",
-    "_NET_CLIENT_LIST",
-    "_NET_CURRENT_DESKTOP",
-    "_NET_DESKTOP_NAMES",
-    "_NET_NUMBER_OF_DESKTOPS",
-    "_NET_SUPPORTED",
-    "_NET_SUPPORTING_WM_CHECK",
-    "_NET_SYSTEM_TRAY_OPCODE",
-    "_NET_SYSTEM_TRAY_ORIENTATION",
-    "_NET_SYSTEM_TRAY_ORIENTATION_HORZ",
-    "_NET_SYSTEM_TRAY_S0",
-    "_NET_WM_DESKTOP",
-    "_NET_WM_NAME",
-    "_NET_WM_STATE",
-    "_NET_WM_STATE_FULLSCREEN",
-    "_NET_WM_WINDOW_TYPE",
-    "_XEMBED",
-    "_XEMBED_INFO",
-    // window types
-    "_NET_WM_WINDOW_TYPE_DESKTOP",
-    "_NET_WM_WINDOW_TYPE_DOCK",
-    "_NET_WM_WINDOW_TYPE_TOOLBAR",
-    "_NET_WM_WINDOW_TYPE_MENU",
-    "_NET_WM_WINDOW_TYPE_UTILITY",
-    "_NET_WM_WINDOW_TYPE_SPLASH",
-    "_NET_WM_WINDOW_TYPE_DIALOG",
-    "_NET_WM_WINDOW_TYPE_DROPDOWN_MENU",
-    "_NET_WM_WINDOW_TYPE_POPUP_MENU",
-    "_NET_WM_WINDOW_TYPE_NOTIFICATION",
-    "_NET_WM_WINDOW_TYPE_COMBO",
-    "_NET_WM_WINDOW_TYPE_DND",
-    "_NET_WM_WINDOW_TYPE_NORMAL",
-];
+gen_enum_with_slice!(
+    /**
+     * Enum with X Atoms. See 'WindowType' for '_NET_WINDOW_TYPE' atoms.
+     */
+    #[derive(Debug, Clone, Hash, PartialEq, Eq)]
+    Atom,
+    /**
+     * A slice containing all enum variants of `Atom`
+     */
+    ATOMS,
+{
+    [Manager, "MANAGER"],
+    [UTF8String, "UTF8_STRING"],
+    [WmClass, "WM_CLASS"],
+    [WmDeleteWindow, "WM_DELETE_WINDOW"],
+    [WmProtocols, "WM_PROTOCOLS"],
+    [WmState, "WM_STATE"],
+    [WmName, "WM_NAME"],
+    [WmTakeFocus, "WM_TAKE_FOCUS"],
+    [NetActiveWindow, "_NET_ACTIVE_WINDOW"],
+    [NetClientList, "_NET_CLIENT_LIST"],
+    [NetCurrentDesktop, "_NET_CURRENT_DESKTOP"],
+    [NetDesktopNames, "_NET_DESKTOP_NAMES"],
+    [NetNumberOfDesktops, "_NET_NUMBER_OF_DESKTOPS"],
+    [NetSupported, "_NET_SUPPORTED"],
+    [NetSupportingWmCheck, "_NET_SUPPORTING_WM_CHECK"],
+    [NetSystemTrayOpcode, "_NET_SYSTEM_TRAY_OPCODE"],
+    [NetSystemTrayOrientation, "_NET_SYSTEM_TRAY_ORIENTATION"],
+    [NetSystemTrayOrientationHorz, "_NET_SYSTEM_TRAY_ORIENTATION_HORZ"],
+    [NetSystemTrayS0, "_NET_SYSTEM_TRAY_S0"],
+    [NetWmDesktop, "_NET_WM_DESKTOP"],
+    [NetWmName, "_NET_WM_NAME"],
+    [NetWmState, "_NET_WM_STATE"],
+    [NetWmStateFullscreen, "_NET_WM_STATE_FULLSCREEN"],
+    [NetWmWindowType, "_NET_WM_WINDOW_TYPE"],
+    [XEmbed, "_XEMBED"],
+    [XEmbedInfo, "_XEMBED_INFO"],
+});
+
+gen_enum_with_slice!(
```

Comment:
Making these atoms a different type feels like it is going to cause some issues down the line. Atoms are frequently used as arguments to XCB API calls and some of the more generic methods may have issues with this change. Please keep them as a single type: sorry I missed the fact that you had done this in your comment on the original issue.

Reply:
At first it looked like a good idea, but when I was changing other parts of the code, I realised it might not have been the best choice. So I agree that it wasn't the best idea.